File: Graphics/Aditya_Rajput/4.py
import numpy as np

v = np.array
l = lambda x : np.linalg.norm(x)

class Ray:    
    def __init__(self, o, d):
        self.origin = o
        self.direction = d

    def __getitem__(self, t):
        return self.origin + t*self.direction

def color(ray):
    if hit_sphere(v([0.0, 0, -1]), 0.5, ray):
        return v([1.0, 0, 0])
    u = ray.direction/np.linalg.norm(ray.direction)
    t = (u[1]+1)/2
    return (1-t)*v([1.0, 1, 1]) + t*v([0.5, 0.7, 1])

def hit_sphere(center, radius, ray):
    AO = ray.origin - center
    a = l(ray.direction)**2
    b = 2*AO.dot(ray.direction)
    c = l(AO)**2 - radius**2
    return b**2 > 4*a*c



nx = 200
ny = 100

with open('4.ppm', 'w') as f:
    print('P3\n{0} {1}\n255\n'.format(nx, ny), file = f)
    
    llc = v([-2.0, -1, -1])
    hori = v([4.0, 0, 0])
    vert = v([0, 2.0, 0])
    orig = v([0.0, 0, 0])
    
    for j in range(ny - 1, -1, -1):
        for i in range(0, nx, 1):
            ray = Ray(orig, llc + (i/nx)*hori + (j/ny)*vert)
            col = color(ray) * 255.99
            print('{0} {1} {2}'.format(col[0], col[1], col[2]), file = f)

print('END')

File: Graphics/Sarthak_Gupta/4.cpp
#include<fstream>
#include "ray.h"

bool hit_sphere(const vec3& center , float radius , const ray& r)
{
    vec3 oc = r.origion() - center;
	float a = dot(r.direction() , r.direction());
	float b = 2.0 * dot(oc , r.direction());
	float c = dot(oc , oc) - radius*radius;
	float discriminant = b*b - 4.0*a*c;
	return (discriminant > 0);
}

vec3 color(const ray& r) {
	if (hit_sphere(vec3(0.0 , 0.0 ,-1) , 0.5 , r)){
		return vec3(1.0 , 0.0 , 0.0);
	} 
//	vec3 unit_direction = unit_vector(r.direction());
//	float yd = 0.5*(unit_direction.y() + 1.0);
//	return (1.0 - yd)*vec3(1.0 , 1.0 , 1.0) + yd*vec3(0.5 , 0.7 , 1.0);
    return vec3(1.0 , 1.0 , 1.0);
}

int main(){
	int npx = 200;
	int npy = 100;
	ofstream ofs("Fourth.ppm" , ios::out | ios::app);
	ofs << "P3\n" << npx << " " << npy << "\n255" <<endl;
	vec3 bottom_left_corner(-2.0 , -1.0 , -1.0);
	vec3 x_motion(4.0 , 0.0 , 0.0);
	vec3 y_motion(0.0 , 2.0 , 0.0);
	vec3 origin(0.0 , 0.0 , 0.0);
	
	for(int j = npy-1 ; j >= 0 ; j--)
	{
		for(int i = 0 ; i < npx ; i++)
		{
			float u = float(i)/float(npx);
			float v = float(j)/float(npy);
			ray r(origin , bottom_left_corner + u*x_motion + v*y_motion);
			vec3 col = color(r).toColor();
			ofs << col.e[0] << " " << col.e[1] << " " << col.e[2] << endl;
		}
	}
	
	ofs.close();
}

/* Made a function whihc checks if there it hits a sphere at hte center and if it does then make that ray red. and rest is same*/

